lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1557,17 +1557,29 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
 
     const bool canEmitIncrementalInfoIntoModule =
         !serializationOpts.DisableCrossModuleIncrementalInfo &&
-        (Action == FrontendOptions::ActionType::MergeModules);
+        (Action == FrontendOptions::ActionType::MergeModules ||
+         Action == FrontendOptions::ActionType::EmitModuleOnly);
     if (canEmitIncrementalInfoIntoModule) {
       const auto alsoEmitDotFile =
           Instance.getInvocation()
               .getLangOptions()
               .EmitFineGrainedDependencySourcefileDotFiles;
 
       using SourceFileDepGraph = fine_grained_dependencies::SourceFileDepGraph;
-      auto *Mod = MSF.get<ModuleDecl *>();
+
+      StringRef fileFingerprint;
+      if (auto Mod = MSF.dyn_cast<ModuleDecl *>()) {
```

Comment:
I'm a bit confused as to how we see SourceFiles here. There shouldn't be any primaries, so this path should only be fed by ModuleDecls, right?

Reply:
Emit module only has source file inputs that end up here. Do you think that shouldn't happen?

I wouldn't expect this job to be passed the primary flag but I could double check in the tests where this failed.

Reply:
My expectation is that separate module mode behaves like WMO, not like single-file mode. There shouldn't be primaries here.

Reply:
Good point! IIUC a jobs with the action `-emit-module` and using a primary-file would be used to create a partial swiftmodule file, so I'll modify this code to avoid jobs with a SourceFile and it should just generate the graph for module-wide emit-module jobs.